xds/internal/balancer/outlierdetection/balancer.go

```diff
@@ -105,3 +132,760 @@ func (bb) ParseConfig(s json.RawMessage) (serviceconfig.LoadBalancingConfig, err
+			if obj == nil { // uneject unconditionally because could have come from an ejected address
+				scw.eject()
+			} else {
+				if obj.latestEjectionTimestamp.IsZero() { // relay new updated subconn state
+					scw.uneject()
+				} else {
+					scw.eject()
+				}
+			}
+		} else { // single address to multiple addresses
+			// 1. Remove Subchannel from Addresses map entry if present in Addresses map.
+			b.removeSubConnFromAddressesMapEntry(scw)
+			// 2. Clear the Subchannel wrapper's Call Counter entry.
+			obj := (*addressInfo)(atomic.LoadPointer(&scw.addressInfo))
+			if obj != nil {
+				obj.callCounter.clear()
+			}
+			// 3. Uneject the Subchannel in case it was previously ejected.
+			scw.uneject()
+		}
+	} else {
+		if len(addrs) == 1 { // multiple addresses to single address
+			// 1. Add Subchannel to Addresses map entry if new address present in map.
+			obj := b.appendIfPresent(addrs[0].Addr, scw)
+			if obj != nil && !obj.latestEjectionTimestamp.IsZero() {
+				scw.eject()
+			}
+		} // else is multiple to multiple - no op, continued to be ignored by outlier detection.
+	}
+
+	scw.addresses = addrs
+}
+
+func (b *outlierDetectionBalancer) ResolveNow(opts resolver.ResolveNowOptions) {
+	b.cc.ResolveNow(opts)
+}
+
+func (b *outlierDetectionBalancer) Target() string {
+	return b.cc.Target()
+}
+
+func max(x, y int64) int64 {
+	if x < y {
+		return y
+	}
+	return x
+}
+
+func min(x, y int64) int64 {
+	if x < y {
+		return x
+	}
+	return y
+}
+
+func (b *outlierDetectionBalancer) intervalTimerAlgorithm() {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	b.timerStartTime = time.Now()
+
+	// 2. For each address, swap the call counter's buckets in that address's
+	// map entry.
+	for _, obj := range b.addrs {
+		obj.callCounter.swap()
+	}
+
+	// 3. If the success_rate_ejection configuration field is set, run the
+	// success rate algorithm.
+	if b.cfg.SuccessRateEjection != nil {
+		b.successRateAlgorithm()
+	}
+
+	// 4. If the failure_percentage_ejection configuration field is set, run the
+	// failure percentage algorithm.
+	if b.cfg.FailurePercentageEjection != nil {
+		b.failurePercentageAlgorithm()
+	}
+
+	// 5. For each address in the map:
+	for addr, obj := range b.addrs {
+		// If the address is not ejected and the multiplier is greater than 0,
+		// decrease the multiplier by 1.
+		if obj.latestEjectionTimestamp.IsZero() && obj.ejectionTimeMultiplier > 0 {
+			obj.ejectionTimeMultiplier--
+			continue
+		}
+		// If the address is ejected, and the current time is after
+		// ejection_timestamp + min(base_ejection_time (type: time.Time) *
+		// multiplier (type: int), max(base_ejection_time (type: time.Time),
+		// max_ejection_time (type: time.Time))), un-eject the address.
+		if !obj.latestEjectionTimestamp.IsZero() && now().After(obj.latestEjectionTimestamp.Add(time.Duration(min(b.cfg.BaseEjectionTime.Nanoseconds()*obj.ejectionTimeMultiplier, max(b.cfg.BaseEjectionTime.Nanoseconds(), b.cfg.MaxEjectionTime.Nanoseconds()))))) {
```

Comment:
Even with the use of max and min, this is still too long to follow.  Please simplify into multiple statements or temporary variables if possible.